File: apps/web/src/types/accounting.ts
// Accounting Types and Interfaces

export enum AccountType {
  ASSET = 'asset',
  LIABILITY = 'liability',
  EQUITY = 'equity',
  REVENUE = 'revenue',
  EXPENSE = 'expense'
}

export enum AccountSubType {
  // Assets
  CURRENT_ASSET = 'current_asset',
  FIXED_ASSET = 'fixed_asset',
  INTANGIBLE_ASSET = 'intangible_asset',
  
  // Liabilities
  CURRENT_LIABILITY = 'current_liability',
  LONG_TERM_LIABILITY = 'long_term_liability',
  
  // Equity
  OWNER_EQUITY = 'owner_equity',
  RETAINED_EARNINGS = 'retained_earnings',
  
  // Revenue
  OPERATING_REVENUE = 'operating_revenue',
  NON_OPERATING_REVENUE = 'non_operating_revenue',
  
  // Expenses
  OPERATING_EXPENSE = 'operating_expense',
  NON_OPERATING_EXPENSE = 'non_operating_expense',
  COST_OF_GOODS_SOLD = 'cost_of_goods_sold'
}

export interface Account {
  id: string
  code: string
  name: string
  type: AccountType
  subType: AccountSubType
  parentId?: string
  description?: string
  balance: number
  isActive: boolean
  createdAt: string
  updatedAt: string
}

export enum TransactionType {
  JOURNAL_ENTRY = 'journal_entry',
  INVOICE = 'invoice',
  PAYMENT = 'payment',
  RECEIPT = 'receipt',
  ADJUSTMENT = 'adjustment'
}

export interface JournalEntry {
  id: string
  entryNumber: string
  date: string
  description: string
  reference?: string
  type: TransactionType
  totalDebit: number
  totalCredit: number
  isPosted: boolean
  createdBy: string
  createdAt: string
  updatedAt: string
  lineItems: JournalLineItem[]
}

export interface JournalLineItem {
  id: string
  journalEntryId: string
  accountId: string
  account?: Account
  description?: string
  debitAmount: number
  creditAmount: number
  sortOrder: number
}

export interface FinancialReport {
  id: string
  name: string
  type: ReportType
  period: ReportPeriod
  startDate: string
  endDate: string
  data: any
  generatedAt: string
  generatedBy: string
}

export enum ReportType {
  BALANCE_SHEET = 'balance_sheet',
  INCOME_STATEMENT = 'income_statement',
  CASH_FLOW = 'cash_flow',
  TRIAL_BALANCE = 'trial_balance',
  GENERAL_LEDGER = 'general_ledger'
}

export enum ReportPeriod {
  MONTHLY = 'monthly',
  QUARTERLY = 'quarterly',
  YEARLY = 'yearly',
  CUSTOM = 'custom'
}

export interface TrialBalance {
  accounts: TrialBalanceAccount[]
  totalDebits: number
  totalCredits: number
  isBalanced: boolean
  asOfDate: string
}

export interface TrialBalanceAccount {
  accountId: string
  accountCode: string
  accountName: string
  accountType: AccountType
  debitBalance: number
  creditBalance: number
}

export interface BalanceSheet {
  asOfDate: string
  assets: BalanceSheetSection
  liabilities: BalanceSheetSection
  equity: BalanceSheetSection
  totalAssets: number
  totalLiabilitiesAndEquity: number
  isBalanced: boolean
}

export interface BalanceSheetSection {
  title: string
  accounts: BalanceSheetAccount[]
  total: number
  subsections?: BalanceSheetSection[]
}

export interface BalanceSheetAccount {
  accountId: string
  accountCode: string
  accountName: string
  balance: number
}

export interface IncomeStatement {
  startDate: string
  endDate: string
  revenue: IncomeStatementSection
  expenses: IncomeStatementSection
  grossProfit: number
  operatingIncome: number
  netIncome: number
}

export interface IncomeStatementSection {
  title: string
  accounts: IncomeStatementAccount[]
  total: number
  subsections?: IncomeStatementSection[]
}

export interface IncomeStatementAccount {
  accountId: string
  accountCode: string
  accountName: string
  amount: number
}

export interface CashFlow {
  startDate: string
  endDate: string
  operatingActivities: CashFlowSection
  investingActivities: CashFlowSection
  financingActivities: CashFlowSection
  netCashFlow: number
  beginningCash: number
  endingCash: number
}

export interface CashFlowSection {
  title: string
  items: CashFlowItem[]
  total: number
}

export interface CashFlowItem {
  description: string
  amount: number
}

// Form interfaces
export interface AccountFormData {
  code: string
  name: string
  type: AccountType
  subType: AccountSubType
  parentId?: string
  description?: string
}

export interface JournalEntryFormData {
  date: string
  description: string
  reference?: string
  type: TransactionType
  lineItems: JournalLineItemFormData[]
}

export interface JournalLineItemFormData {
  accountId: string
  description?: string
  debitAmount: number
  creditAmount: number
}

// API Response interfaces
export interface AccountsResponse {
  accounts: Account[]
  total: number
  page: number
  limit: number
}

export interface JournalEntriesResponse {
  entries: JournalEntry[]
  total: number
  page: number
  limit: number
}

// Utility functions
export const getAccountTypeDisplayName = (type: AccountType): string => {
  const names = {
    [AccountType.ASSET]: 'Asset',
    [AccountType.LIABILITY]: 'Liability',
    [AccountType.EQUITY]: 'Equity',
    [AccountType.REVENUE]: 'Revenue',
    [AccountType.EXPENSE]: 'Expense'
  }
  return names[type]
}

export const getAccountSubTypeDisplayName = (subType: AccountSubType): string => {
  const names = {
    [AccountSubType.CURRENT_ASSET]: 'Current Asset',
    [AccountSubType.FIXED_ASSET]: 'Fixed Asset',
    [AccountSubType.INTANGIBLE_ASSET]: 'Intangible Asset',
    [AccountSubType.CURRENT_LIABILITY]: 'Current Liability',
    [AccountSubType.LONG_TERM_LIABILITY]: 'Long-term Liability',
    [AccountSubType.OWNER_EQUITY]: 'Owner Equity',
    [AccountSubType.RETAINED_EARNINGS]: 'Retained Earnings',
    [AccountSubType.OPERATING_REVENUE]: 'Operating Revenue',
    [AccountSubType.NON_OPERATING_REVENUE]: 'Non-operating Revenue',
    [AccountSubType.OPERATING_EXPENSE]: 'Operating Expense',
    [AccountSubType.NON_OPERATING_EXPENSE]: 'Non-operating Expense',
    [AccountSubType.COST_OF_GOODS_SOLD]: 'Cost of Goods Sold'
  }
  return names[subType]
}

export const getTransactionTypeDisplayName = (type: TransactionType): string => {
  const names = {
    [TransactionType.JOURNAL_ENTRY]: 'Journal Entry',
    [TransactionType.INVOICE]: 'Invoice',
    [TransactionType.PAYMENT]: 'Payment',
    [TransactionType.RECEIPT]: 'Receipt',
    [TransactionType.ADJUSTMENT]: 'Adjustment'
  }
  return names[type]
}

export const formatCurrency = (amount: number): string => {
  return new Intl.NumberFormat('id-ID', {
    style: 'currency',
    currency: 'IDR',
    minimumFractionDigits: 0,
    maximumFractionDigits: 0
  }).format(amount)
}

export const isDebitAccount = (accountType: AccountType): boolean => {
  return [AccountType.ASSET, AccountType.EXPENSE].includes(accountType)
}

export const isCreditAccount = (accountType: AccountType): boolean => {
  return [AccountType.LIABILITY, AccountType.EQUITY, AccountType.REVENUE].includes(accountType)
}

File: crates/shared-types/src/accounting.rs
use chrono::{DateTime, NaiveDate, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct Account {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub code: String,
    pub name: String,
    pub account_type: AccountType,
    pub account_subtype: Option<String>,
    pub parent_id: Option<Uuid>,
    pub is_active: bool,
    pub description: Option<String>,
    pub balance_type: BalanceType,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum AccountType {
    Asset,
    Liability,
    Equity,
    Revenue,
    Expense,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum BalanceType {
    Debit,
    Credit,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct JournalEntry {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub entry_number: String,
    pub entry_date: NaiveDate,
    pub reference: Option<String>,
    pub description: String,
    pub total_debit: Decimal,
    pub total_credit: Decimal,
    pub status: JournalEntryStatus,
    pub created_by: Uuid,
    pub posted_by: Option<Uuid>,
    pub posted_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub lines: Option<Vec<JournalEntryLine>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum JournalEntryStatus {
    Draft,
    Posted,
    Reversed,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct JournalEntryLine {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub journal_entry_id: Uuid,
    pub account_id: Uuid,
    pub account: Option<Account>,
    pub description: Option<String>,
    pub debit_amount: Decimal,
    pub credit_amount: Decimal,
    pub line_number: i32,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct AccountBalance {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub account_id: Uuid,
    pub account: Option<Account>,
    pub period_year: i32,
    pub period_month: i32,
    pub opening_balance: Decimal,
    pub debit_total: Decimal,
    pub credit_total: Decimal,
    pub closing_balance: Decimal,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct FinancialPeriod {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub name: String,
    pub start_date: NaiveDate,
    pub end_date: NaiveDate,
    pub status: PeriodStatus,
    pub is_current: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum PeriodStatus {
    Open,
    Closed,
}

// Request/Response DTOs
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct CreateAccountRequest {
    pub code: String,
    pub name: String,
    pub account_type: AccountType,
    pub account_subtype: Option<String>,
    pub parent_id: Option<Uuid>,
    pub description: Option<String>,
    pub balance_type: BalanceType,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct UpdateAccountRequest {
    pub code: Option<String>,
    pub name: Option<String>,
    pub account_type: Option<AccountType>,
    pub account_subtype: Option<String>,
    pub parent_id: Option<Uuid>,
    pub description: Option<String>,
    pub balance_type: Option<BalanceType>,
    pub is_active: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct CreateJournalEntryRequest {
    pub entry_date: NaiveDate,
    pub reference: Option<String>,
    pub description: String,
    pub lines: Vec<CreateJournalEntryLineRequest>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct CreateJournalEntryLineRequest {
    pub account_id: Uuid,
    pub description: Option<String>,
    pub debit_amount: Decimal,
    pub credit_amount: Decimal,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct UpdateJournalEntryRequest {
    pub entry_date: Option<NaiveDate>,
    pub reference: Option<String>,
    pub description: Option<String>,
    pub lines: Option<Vec<CreateJournalEntryLineRequest>>,
}

// Financial Reports
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct TrialBalance {
    pub period_start: NaiveDate,
    pub period_end: NaiveDate,
    pub accounts: Vec<TrialBalanceAccount>,
    pub total_debits: Decimal,
    pub total_credits: Decimal,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct TrialBalanceAccount {
    pub account: Account,
    pub debit_balance: Decimal,
    pub credit_balance: Decimal,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct BalanceSheet {
    pub as_of_date: NaiveDate,
    pub assets: Vec<BalanceSheetSection>,
    pub liabilities: Vec<BalanceSheetSection>,
    pub equity: Vec<BalanceSheetSection>,
    pub total_assets: Decimal,
    pub total_liabilities: Decimal,
    pub total_equity: Decimal,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct BalanceSheetSection {
    pub section_name: String,
    pub accounts: Vec<BalanceSheetAccount>,
    pub section_total: Decimal,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct BalanceSheetAccount {
    pub account: Account,
    pub balance: Decimal,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct IncomeStatement {
    pub period_start: NaiveDate,
    pub period_end: NaiveDate,
    pub revenue: Vec<IncomeStatementSection>,
    pub expenses: Vec<IncomeStatementSection>,
    pub total_revenue: Decimal,
    pub total_expenses: Decimal,
    pub net_income: Decimal,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct IncomeStatementSection {
    pub section_name: String,
    pub accounts: Vec<IncomeStatementAccount>,
    pub section_total: Decimal,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct IncomeStatementAccount {
    pub account: Account,
    pub amount: Decimal,
}
